library/data_structures/seg_tree_uncommon/find_first.hpp

```diff
@@ -0,0 +1,31 @@
+/** @file */
+#pragma once
+/**
+ * @param le,ri defines range [le, ri)
+ * @param f(tree[u], tl, tr) defines a function that returns true if the subtree
+ * contains an element that satisfies the condition
+ * @returns the index of the first element in the range that satisfies the
+ * condition described in `f`, if no such element exists then `ri` is returned
+ * @time O(log(n))
+ * @space O(log(n)) for recursion stack; no new nodes are allocated
+ */
+int find_first(int le, int ri, function<bool(int64_t, int, int)> f) {
```

Comment:
it might be cleaner to templatize `f`

Reply:
It ended up being cleaner & faster, so I updated this

see
without template: https://codeforces.com/contest/1705/submission/250938915
with template: https://codeforces.com/contest/1705/submission/250940123